File: inst/Matlab/fdaM/fbplot.m
   
function [depth, outpoint]=fbplot(data,x,depth,method,show,prob,color,outliercol,barcol,fullout,factor)
% Produces functional boxplots or enhanced functional boxplots of the given functional data. 
% It can also be used to carry out functional data ordering based on band depth. 
%
%Arguments
%
%	data: a p-by-n functional data matrix where n is the number of curves, and p is defined below.
%         alternatively, a functional data object or functional parameter
%         object can also be specified. 
%	x:	the x coordinates of curves. Defaults to 1:p where p is the number of x coordinates.
%	method: the method to be used to compute band depth. Can be one of "BD2", "BD3", "MBD" or "Both" 
%	        with a default of "MBD".
%	depth:	a vector giving band depths of curves. If missing, band depth computation is conducted.
%	show: logical. If TRUE (the default) then a functional boxplot is produced. If not, band depth
%	      and outliers are returned.
%	prob: a vector giving the probabilities of central regions in a decreasing order, then an enhanced 
%	      functional boxplot is produced. Defaults to be 0.5 and a functional boxplot is plotted.
%	color:	a vector giving the colors of central regions from light to dark for an enhanced functional 
%	       boxplot. Defaults to be magenta for a functional boxplot.
%	outliercol: color of outlying curves. Defaults to be red.
%	barcol: color of bars in a functional boxplot. Defaults to be blue.
%	fullout: logical for plotting outlying curves. If FALSE (the default) then only the part outside the
%	         box is plotted. If TRUE, complete outling curves are plotted.
%	factor: the constant factor to inflate the middle box and determine fences for outliers. Defaults to 
%	        be 1.5 as in a classical boxplot.
%
%Details
%
%	For functional data, the band depth (BD) or modifed band depth (MBD) allows for ordering a sample of 
%	curves from the center outwards and, thus, introduces a measure to define functional quantiles and 
%	the centrality or outlyingness of an observation. A smaller rank is associated with a more central 
%	position with respect to the sample curves. "BD2" uses two curves to determine a band and "BD3" uses 
%	three curves. BD usually provides many ties (curves have the same depth values), but MBD does not. 
%	The method "Both" uses BD2 first and then uses MBD to break ties.
%
%Value
%
%	depth:	band depths of given curves.
%	outpoint: column indices of detected outliers.
%
%Author(s)
%
%	Ying Sun: sunwards@stat.tamu.edu
%
%	Marc G. Genton: genton@stat.tamu.edu
%
%References
%
%	Sun, Y. and Genton, M. G. (2011), "Functional Boxplots," Journal of Computational and Graphical Statistics,
%	to appear.
%
%	Lopez-Pintado, S. and Romo, J. (2009), "On the concept of depth for functional data," Journal of the American
%	Statistical Association, 104, 718-734.
%

% %%default values
% 
% factor=1.5; 
% fullout='False'; 
% barcol='b'; 
% outliercol='r'; 
% color='m'; 
% prob=0.5; 
% show='True';
% method='MBD'; 
% depth=[]; 

% Examples
%
% 	clear all;
%
% 	ncasem = 39;
% 	ncasef = 54;
% 	nage   = 31;
% 
% 	fid = fopen('hgtm.dat','rt');
% 	hgtmmat = reshape(fscanf(fid,'%f'),[nage,ncasem]);
% 
% 	fid = fopen('hgtf.dat','rt');
% 	hgtfmat = reshape(fscanf(fid,'%f'),[nage,ncasef]);
% 
% 	age = [ 1:0.25:2, 3:8, 8.5:0.5:18 ]';
%
%%fbplot of boys' height
%	fbplot(hgtmmat,age,depth,method,show,prob,color,outliercol,barcol,fullout,factor)
%	xlim([0.5,18.5])
%	ylim([60,200])
%	xlabel('Age (Years)')
%	ylabel('Height (cm)')
%	title('Boys')
%
%%fbplot of girls' height
%	fbplot(hgtfmat,age,depth,method,show,prob,color,outliercol,barcol,fullout,factor)
%	xlim([0.5,18.5])
%	ylim([60,200])
%	xlabel('Age (Years)')
%	ylabel('Height (cm)')
%	title('Girls')
 
function combinat=combinat(n,p)
if n<p 
combinat=0;
else
   combinat=nchoosek(n,p);
end
end

function resultados=estaEntre(v,matrizDatos)
% The input in this function is the data matrix and the pair of indexes and the output is a vector of dimension n, where the coordinate k takes the value 1 or 0 whether the observation k is inside the band or not.
[n,p]=size(matrizDatos);                      
Z=matrizDatos;
inf=min(Z(v,:))';
sup=max(Z(v,:))';
resultados=sum((and(Z'<=sup*ones(1,n),Z'>=inf*ones(1,n))))==p;
end

function [contg]=BD2(matrizDatos)
% With this function we calculate the band depth of every observation in the matrix of data: matrizDatos. The rows are the observations and the columns are the variables.
tic
[n,p]=size(matrizDatos);
cont=zeros(1,n); % The depths are initialized to 0.
for i=1:((n+2)/3)
   for j=(i+1):(n) % We choose pairs of indexes in all the possible ways.
      cont=cont+estaEntre([i j],matrizDatos); % With this subfuntion we check whether the observations are inside the band defined by observation i and j.                 
   end
end	
   contg=cont/combinat(n,2);
toc
end

function [contg]=BD3(matrizDatos)
% With this function we calculate the band depth with J=3. The imput is the data matrix, where the rows are the observations and the columns the variables. 
tic
[n,p]=size(matrizDatos); 
cont=zeros(1,n);      % Initialize the depth of each observation to zero.
                      % Select three observations from the sample in all the possible ways. 
for i=1:(n-2)
   for j=(i+1):(n-1)
      for k=(j+1):n
         cont=cont+estaEntre([i j k],matrizDatos);   % In this subfunction we check which observations from the sample is inside the band delimeted by observations i,j and k.           
         
      end
   end
end	
contg=cont/combinat(n,3);
toc
end

function resultado=a(v,matrizDatos) 
[n,p]=size(matrizDatos);
Z=matrizDatos;
inf=(min(Z(v,:)))';
sup=(max(Z(v,:)))';
resul=sum((and(Z'<=sup*ones(1,n),Z'>=inf*ones(1,n))));
resultado=(resul/p); % Proportion of coordinates of each observation from the sample that is inside the band delimited by pairs v of functions from the sample.
end

function [contg]=MBD(matrizDatos)
% This function calculates the generalized band depth of a set of data
tic
[n,p]=size(matrizDatos); % size of the data matrix
cont=zeros(1,n); % The initial value of the depth of the observations from the matrix is zero
for i=1:(n-1)
    for j=(i+1):(n) % consider all possible pairs of functions
        
        cont=cont+a([i j],matrizDatos);  % In cont we save the generalized band depth of each observation (function)
        
    end
end
contg=cont;
contg=cont/combinat(n,2);
toc
end

%%default values

if nargin<11, 	factor=1.5;  end
if nargin<10, 	fullout='False'; end
if nargin<9, 	barcol='b'; end
if nargin<8, 	outliercol='r'; end
if nargin<7, 	color='m'; end
if nargin<6, 	prob=0.5; end
if nargin<5, 	show='True'; end
if nargin<4, 	method='MBD'; end
if nargin<3, 	depth=[]; end
if nargin<2,    x=[]; end

% If data is an fd object or fdPar object extract it

if(isa_fdPar(data)), data = getfd(data); end
if(isa_fd(data))
   if isempty(x) || length(x)==1,
      rr = getbasisrange(getbasis(data));
      if isempty(x), npt = 101;
      else npt = x; end
      x = linspace(rr(1),rr(2),npt)';
   end
   data = eval_fd(x,data);
end

% default value for x

[tp,n]=size(data);
if isempty(x), x = (1:tp)'; end
if (length(x) ~= tp), error('Dimensions of data and x do not match'); end


  %compute band depth	
  if isempty(depth)
	if strcmp(method,'BD2')
        depth=BD2(data');
    elseif strcmp(method,'BD3')
            depth=BD3(data');
    elseif strcmp(method,'MBD')
            depth=MBD(data');
    elseif strcmp(method,'Both')
            depth=round(BD2(data)*10000)+MBD(data');  
	end
  end
    
	[dp_s,index]=sort(depth,'descend');
    for pp=1:length(prob)
		m=ceil(n*prob(pp));%at least 50%
		center=data(:,index(1:m));
		out=data(:,index((m+1):n));
		inf=min(center,[],2)';
		sup=max(center,[],2)';
		if prob(pp)==0.5 %check outliers
			dist=factor*(sup-inf);
			upper=sup+dist;
			lower=inf-dist;
			%outlier column
			outly=sum(or(data<=lower'*ones(1,n),data>=upper'*ones(1,n)));
			outpoint=find(outly);
			out=data(:,outpoint);
            good=data;
            good(:,outpoint)=[];
			maxcurve=max(good,[],2)';
			mincurve=min(good,[],2)';
			if sum(outly)>0
				if show 
				plot(x,out,'--r');
				end
			end
			barval=(x(1)+x(tp))/2;
            loc=find(sort([x;barval])==barval);
			bar=loc(1);
			if show
			hold all;
			line([x(bar) x(bar)],[maxcurve(bar) sup(bar)],'Color',barcol,'LineWidth',2);
			hold all;
		    line([x(bar) x(bar)],[mincurve(bar) inf(bar)],'Color',barcol,'LineWidth',2);
			end
		end
		
		if show 
			hold all;
            [xinv,xindex]=sort(x,'descend');
            xx=[x;xinv];
    		supinv=sup(xindex);
            yy=[inf,supinv];
			h=fill(xx,yy,color(pp));
			if prob(pp)==0.5
			set(h,'edgecolor',barcol,'LineWidth',2);
			else 
			set(h,'edgecolor',NA);
			end
		end
		if show
			hold all;
			plot(x,data(:,index(1)),'Color','k','LineWidth',2);
			plot(x,maxcurve,'Color','b','LineWidth',2);
			plot(x,mincurve,'Color','b','LineWidth',2);
			if fullout
				if sum(outly)>0 
					hold all;
					plot(x,out,'Color',outliercol);
				end
			end
		end
    end
    hold off;
depth
outpoint
end

File: R/fbplot.R

#BD2
BD2=function(matrizDatos){
	n=dim(matrizDatos)[1]
	p=dim(matrizDatos)[2]
	cont=rep(0,n)
	for (i in 1:(n-1)){
		for (j in (i+1):n){
			cont=cont+estaEntre(c(i,j),matrizDatos)
		}
	}
	contg=(cont/combinat(n,2))
}

#indicator function
estaEntre=function(v,matrizDatos){
	n=dim(matrizDatos)[1]
	p=dim(matrizDatos)[2]
	Z=matrizDatos
	inf=t(apply(Z[v,],2,min))
	sup=t(apply(Z[v,],2,max))
	resultados=colSums((t(Z)<=t(sup)%*%rep(1,n))* (t(Z)>=t(inf)%*%rep(1,n)))==p
}

#combination
combinat=function(n,p){
	if (n<p){combinat=0}
	else {combinat=factorial(n)/(factorial(p)*factorial(n-p))}
}

#BD3
BD3=function(matrizDatos){
	n=dim(matrizDatos)[1]
	p=dim(matrizDatos)[2]
	cont=rep(0,n)
	for (i in 1:(n-2)){
		for (j in (i+1):(n-1)){
			for (k in (j+1):n){
				cont=cont+estaEntre(c(i,j,k),matrizDatos)
			}
		}
	}
	contg=(cont/combinat(n,3))
}

#MBD
MBD=function(matrizDatos){
	n=dim(matrizDatos)[1]
	p=dim(matrizDatos)[2]
	cont=rep(0,n)
	for (i in 1:(n-1)){
		for (j in (i+1):n){
			cont=cont+aprops(c(i,j),matrizDatos)
		}
	}
	contg=(cont/combinat(n,2))
}

#proportion function
aprops=function(v,matrizDatos){
	n=dim(matrizDatos)[1]
	p=dim(matrizDatos)[2]
  Z=matrizDatos
	inf=t(apply(Z[v,],2,min))
	sup=t(apply(Z[v,],2,max))
	resul=colSums((t(Z)<=t(sup)%*%rep(1,n))* (t(Z)>=t(inf)%*%rep(1,n)))
	resultado=(resul/p)
}




#function boxplot
#fit: p by n functional data matrix, n is the number of curves
#method: BD2, BD3, MBD
fbplot=function(fit,x=NULL,method='MBD',depth=NULL,plot=TRUE,prob=0.5,color=6,outliercol=2,
				barcol=4,fullout=FALSE, factor=1.5,...){
				
  if(is.fdSmooth(fit) | is.fdPar(fit)){ fit = fit$fd }  
	if(is.fd(fit)){
    if(length(x)==0){
      x = seq(fit$basis$rangeval[1],fit$basis$rangeval[2],len=101)
    }
    fit = eval.fd(x,fit)
  }				
				
	tp=dim(fit)[1]
	n=dim(fit)[2]
	if (length(x)==0) {x=1:tp}
  #compute band depth	
  if (length(depth)==0){
	if (method=='BD2') {depth=BD2(t(fit))}
	else if (method=='BD3') {depth=BD3(t(fit))}
	else if (method=='MBD') {depth=MBD(t(fit))}
	else if (method=='Both') {depth=round(BD2(t(fit)),4)*10000+MBD(t(fit))}
  }

	dp_s=sort(depth,decreasing=T)
	index=order(depth,decreasing=T)
	if (plot) {
	plot(x,fit[,index[1]],lty=1,lwd=2,col=1,type='l',...)
	}
	for (pp in 1:length(prob)){
		m=ceiling(n*prob[pp])#at least 50%
		center=fit[,index[1:m]]
		out=fit[,index[(m+1):n]]
		inf=apply(center,1,min)
		sup=apply(center,1,max)
		
		if (prob[pp]==0.5){ #check outliers
			dist=factor*(sup-inf)
			upper=sup+dist
			lower=inf-dist
			outly=(fit<=lower)+(fit>=upper)
			outcol=colSums(outly)
			remove=(outcol>0)
			#outlier column
			colum=1:n
			outpoint=colum[remove==1]
			out=fit[,remove]
			woout=fit
			good=woout[,(remove==0),drop=FALSE]
			maxcurve=apply(good,1,max)
			mincurve=apply(good,1,min)
			if (sum(outly)>0){
				if (plot) {
				matlines(x,out,lty=2,col=outliercol,type='l',...)
				}
			}
			barval=(x[1]+x[tp])/2
			bar=which(sort(c(x,barval))==barval)[1]
			if (plot) {
			lines(c(x[bar],x[bar]),c(maxcurve[bar],sup[bar]),col=barcol,lwd=2)
		    lines(c(x[bar],x[bar]),c(mincurve[bar],inf[bar]),col=barcol,lwd=2)
			}
		}
		xx=c(x,x[order(x,decreasing=T)])
		supinv=sup[order(x,decreasing=T)]
		yy=c(inf,supinv)
		if (plot) {
		if (prob[pp]==0.5) {polygon(xx,yy,col=color[pp],border=barcol,lwd=2)}
		else {polygon(xx,yy,col=color[pp],border=NA)}
		}
	}
	if (plot) {
	lines(x,fit[,index[1]],lty=1,lwd=2,col=1,type='l')
	lines(x,maxcurve,col=barcol,lwd=2)
	lines(x,mincurve,col=barcol,lwd=2)
	if (fullout) {
		if (sum(outly)>0){
				if (plot) {
				matlines(x,out,lty=2,col=outliercol,type='l',...)
				}
			}
		}
	}
	return(list(depth=depth,outpoint=outpoint))
}


